### PR TITLE
docs: Phase 1 source-to-kernel provenance audit

### LINKED ISSUE #2574: Provenance Audit: Map metadata across the compilation pipeline
## Overview

Trace a reference model through every stage of the Spyre compilation pipeline and document where provenance metadata exists, where it transforms, and where it drops. This audit establishes the ground truth for all subsequent implementation work.

## Acceptance Criteria

- [x] Select a reference model (single linear layer or equivalent) and trace it through all six stages:
  1. PyTorch model source
  2. FX graph (pre-grad and post-grad)
  3. Inductor passes
  4. LoopLevelIR
  5. OpSpec
  6. SuperDSC JSON
- [x] At each stage, document:
  - Which provenance fields are present (e.g., `stack_trace`, `original_aten`, `from_node`, `origins`, `origin_node`, `traceback`)
  - The data structure and format of each field
  - How the field was populated (copied, derived, or generated)
  - Whether the field survives to the next stage
- [x] Identify the exact code locations where metadata is dropped or fails to propagate
- [x] Produce a pipeline diagram showing metadata flow (aligned with `docs/source/compiler/architecture.md`)
- [x] Produce an accompanying table: stage × metadata field → present / absent / transformed

## Key Files to Investigate

| Stage | Key files |
|-------|-----------|
| FX metadata | Upstream PyTorch Inductor (node.meta fields) |
| LoopLevelIR | `torch_spyre/_inductor/scratchpad.py` |
| Scheduler metadata | `torch_spyre/_inductor/scheduler.py` (`get_kernel_metadata()`) |
| OpSpec | `torch_spyre/_inductor/op_spec.py` |
| SuperDSC codegen | `torch_spyre/_inductor/codegen/superdsc.py`, `torch_spyre/_inductor/codegen/compute_ops.py` |

## Timeline

May 26 – June 12, 2026 (3 weeks)

## Notes

- Use `TORCH_COMPILE_DEBUG=1` and `TORCH_LOGS="+inductor"` to capture intermediate artifacts
- Compare against upstream Inductor's `inductor_provenance_tracking_node_mappings.json` for field naming conventions
- The audit should answer: "If I add a `debug_handle` at OpSpec, what upstream fields can I derive it from?"

### PR DESCRIPTION
**Pull request by gsmoon97**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [ ] feature
- [x] documentation
- [ ] cleanup

#### What this PR does:

Adds Phase 1 source-to-kernel provenance audit tooling under `docs/source/user_guide/examples/provenance/`. It traces a reference `SimpleMLP` through a single `torch.compile` and records, at each of six pipeline stages, which provenance fields are present on the actual compile-path objects — producing a measurement-only Markdown report plus an interpretation write-up.

Stages: PyTorch source → FX (pre/post-grad) → Inductor passes → LoopLevelIR → OpSpec → SuperDSC JSON. All observation is via read-only monkey-patches installed for the duration of one compile.

#### Which issue(s) this PR is related to:

Fixes #2574
Feeds into #2575 (debug_handle schema)

#### Special notes for your reviewer:

- Requires a Spyre device to run (first device access triggers init, ~60s); no unit-test coverage since it's device-dependent.
- Rebased onto latest `main` and relocated from `examples/provenance/` to `docs/source/user_guide/examples/provenance/` to follow the examples refactor in #2703.

#### Does this PR introduce a user-facing change?

No — read-only tooling, no change to torch-spyre behavior or architecture docs.

#### Additional note:

Findings: provenance survives on the IR `origins` attribute through LoopLevelIR but is dropped at `OpSpec`, so it is absent from the emitted SuperDSC JSON.

### CODE REVIEW

File: docs/source/user_guide/examples/provenance/pipeline/superdsc.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    "traceback",
+    "provenance",
+    "debug_handle",
+    "source",
+]
+
+
+def _scan_json_for_provenance(text: str) -> dict[str, bool]:
+    low = text.lower()
+    return {name: (name.lower() in low) for name in PROVENANCE_FIELD_NAMES}
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/superdsc.py]: The substring scan on raw JSON text will produce false positives. `"source"` matches unrelated keys and values everywhere, and `"origin"` hits any string containing that substring. You already call `json.loads(text)` one line above. Walk the parsed dict's keys instead, check for exact matches, and drop the overly generic `"source"` token.

File: docs/source/user_guide/examples/provenance/pipeline/captures.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+IR_ATTR_FIELDS = ["origins", "origin_node", "traceback"]
+
+_MAX = 200  # repr truncation
+
+
+def _safe(obj: Any) -> str:
+    try:
+        return str(obj)[:_MAX]
+    except Exception as e:  # pragma: no cover - defensive
+        return f"<unreprable {type(obj).__name__}: {e}>"
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/captures.py]: Nit: `"unreprable"` should be `"unrepr-able"` or `"unrepresentable"`.

File: docs/source/user_guide/examples/provenance/pipeline/captures.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    saved: list[tuple[Any, str, Any]] = []  # (obj, attr, original)
+
+    def _patch(obj, attr, factory):
+        original = getattr(obj, attr)
+        setattr(obj, attr, factory(original))
+        saved.append((obj, attr, original))
+
+    # ---- Stage 2 pre-grad ------------------------------------------------
+    def _pre_grad(original):
+        def wrapper(self, graph):
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/captures.py]: The design note about degrading to missing data rather than crashing is good. But the try/except inside the wrapper will not catch a `TypeError` from argument-binding if an upstream signature changes. Python raises that *before* entering your try block. Switching to `def wrapper(self, *args, **kwargs)` and extracting what you need positionally would match the stated design intent. Not urgent since this is pinned to a specific torch build, but worth considering.

File: docs/source/user_guide/examples/provenance/pipeline/report.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
+    pre = capture["stage2_pre_grad"]["nodes"]
+    post = capture["stage2_post_grad"]["nodes"]
+    passes_before = capture["stage3_passes"].get("before", [])
+    looplevel = capture["stage4_looplevel"]["operations"]
+    opspec_ops = capture["stage5_opspec"]["ops"]
+    opspec_fields = capture["stage5_opspec"]["opspec_fields"] or []
+    kernels = capture["stage6_kernels"]["kernels"]
+    sdsc_present = bundles.get("provenance_present", False)
+    sdsc_cell = TICK if sdsc_present else CROSS
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/report.py:159]: The SuperDSC column uses a single global boolean for all IR fields. Today that is correct since provenance is uniformly absent, but if Phase 2 threads even one field through, this will silently misrepresent the others. A TODO noting this simplification only holds while provenance is fully absent from the JSON would help.

File: docs/source/user_guide/examples/provenance/pipeline/report.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+def _ir_value(op: dict, field: str):
+    return op.get("fields", {}).get(field)
+
+
+def _ir_present(op: dict, field: str) -> bool:
+    v = _ir_value(op, field)
+    if field == "origins":
+        return bool(v)
+    if field == "get_stack_traces":
+        return bool(v) and v != "OrderedSet([])"
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/report.py]: Hardcoding `"OrderedSet([])"` as the empty check is brittle. If the repr changes (spacing, parens, import path in the class name), this silently breaks. Can you compare against falsy-ness of the underlying value instead of matching its string representation?

File: docs/source/user_guide/examples/provenance/pipeline/captures.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+            try:
+                name = getattr(op, attr)()
+                break
+            except Exception:
+                pass
+    # Record whether each provenance attribute *exists* on the object, separate
+    # from its value. This distinguishes "the slot exists but is empty" (a real
+    # absence / unused channel -> ✗) from "not an attribute of this object at
+    # all" (not applicable -> ➖). getattr-with-default can't tell these apart.
+    attr_exists = {a: hasattr(op, a) for a in ("origins", "origin_node", "traceback")}
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/pipeline/captures.py]: You are tracking `attr_exists` for the three main IR attrs but `get_stack_traces` is not included. That means `_ir_attr_exists` in report.py defaults it to `True`. Correct today since it is a real method on ComputedBuffer, but recording it explicitly keeps the contract tight if this tooling outlives the current IR shape.

File: docs/source/user_guide/examples/provenance/audit.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+import pathlib
+import sys
+
+# Force-disable Inductor caches before torch is imported anywhere downstream.
+os.environ.setdefault("TORCHINDUCTOR_FORCE_DISABLE_CACHES", "1")
+
+import torch
+import torch_spyre  # noqa: F401 — registers the Spyre backend
+
+sys.path.insert(0, str(pathlib.Path(__file__).parent))
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/provenance/audit.py]: `sys.path.insert(0, ...)` works here but breaks if anyone imports audit.py from a different working directory or runs it inside a larger test harness. Would a relative import or an `__init__.py` at the provenance level be cleaner?

**Review by kaoutar55 — COMMENTED** [review]
The analysis doc is solid. The metadata flow diagram and field-relationship tree make the Phase 2 design decision straightforward.

Question: are the pre-generated `.md` reports meant to stay committed as reference snapshots, or should they be regenerated on each run? If the latter, `.gitignore` them and add a `make report` target or equivalent.

File: docs/source/user_guide/examples/provenance/pipeline/superdsc.py (hunk 1 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/pipeline/superdsc.py]: Thanks for pointing it out. I replaced the raw-text substring scan with an walk over the parsed JSON keys to detect only exact matches, as you suggested. This would prevent any potential false positives from an unrelated key.

File: docs/source/user_guide/examples/provenance/pipeline/captures.py (hunk 3 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/pipeline/captures.py]: Valid point. I switched all the pass wrappers to accept `*args`, `**kwargs` and extract the arguments positionally inside the try block, so an upstream signature change degrades to recorded missing-data instead of raising a `TypeError` before the guard can catch it.

File: docs/source/user_guide/examples/provenance/pipeline/report.py (hunk 4 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/pipeline/report.py:159]: Agreed. I updated the SuperDSC audit script to trace per-field instead of one global flag. It now counts each provenance field across the scanned sdsc_*.json files and renders like the other columns, so a future partial state isn't misrepresented. Please note that since the change is quite minimal, I implemented the changes rather than leaving it as a TODO. For reference, I also did a full re-sweep to ensure that the audit script does a population check (field exists AND value non-empty) for every stage when checking whether a provenance field is dropped or maintained.

File: docs/source/user_guide/examples/provenance/pipeline/captures.py (hunk 6 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/pipeline/captures.py]: Understood. I added `get_stack_traces` to the `attr_exists` record in `_summarize_ir_attrs,` so `_ir_attr_exists` reads its real `hasattr` value instead of just blindly defaulting to `True`. This would keep the contract explicit, if the IR shape were to change later.

File: docs/source/user_guide/examples/provenance/pipeline/report.py (hunk 5 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/pipeline/report.py]: Thanks for pointing it out. I moved `get_stack_traces` emptiness detection to capture time and compare the underlying value's falsiness (a nonempty bool) instead of matching the hard-coded `"OrderedSet([])"` repr.

File: docs/source/user_guide/examples/provenance/audit.py (hunk 7 above)
Comment (gsmoon97) [reply docs/source/user_guide/examples/provenance/audit.py]: Regarding this, I removed the manual `sys.path.insert`, since it was actually redundant. Python already prepends a script's own directory to `sys.path` when you run it directly, so `reference_mlp` and `pipeline` resolve from any working directory, without the global path mutation.

However, I did choose to keep `audit.py` as a run-only entry point for now. Making it importable would require switching to relative imports for all subfiles and running the script via `python -m`, which would break the documented `python audit.py` usage. If we ever need to import the audit itself (e.g. from a test harness), then we can go ahead with the full `__init__.py` + `python -m` change.

**Comment by gsmoon97** [comment]
Hi @kaoutar55, thank you very much for the detailed review. I've pushed updates addressing all your review comments. 

I also made the following changes to keep the audit consistent and clear across all stages:

- Population check standardized across every stage, with per-field / per-instance matrix columns i.e., a provenance field is considered *carried* only if the field key exists **and** its value is non-empty.
- Relabeled the IR matrix columns so they read as one IR before/after the pre-scheduling passes:
  - "Inductor passes" → "LoopLevelIR (pre-pass)"
  - "LoopLevelIR" → "LoopLevelIR (post-pass)"

Please let me know if you have any questions. Thanks!

**Review by kaoutar55 — APPROVED** [review]
LGTM @gsmoon97 . All feedback was properly addressed. The one sys.path item is fine as a run-only-by-design choice; non-blocking. Thanks for the thorough follow-up!

**Review by kaoutar55 — DISMISSED** [review]

**Comment by kaoutar55** [comment]
@gsmoon97 
the Build docs CI check is failing on this PR. Quick fix.
You added three new Markdown pages under docs/source/user_guide/examples/provenance/ (README.md, provenance_audit.md, provenance_analysis.md). Sphinx requires every doc page to be reachable from a toctree, and examples/index.md doesn't have one. So these three are "orphans." CI treats warnings as errors (-W), so the build fails:

provenance/README.md: WARNING: document isn't included in any toctree
provenance/provenance_audit.md: WARNING: document isn't included in any toctree
provenance/provenance_analysis.md: WARNING: document isn't included in any toctree
I suggest this fix: Add this toctree to docs/source/user_guide/examples/index.md (under a new ## Provenance heading is fine):


```{toctree}
:maxdepth: 1
:caption: Provenance Audit

provenance/README
provenance/provenance_audit
provenance/provenance_analysis
```
Paths are relative to index.md, no .md extension. Same pattern as docs/source/user_guide/debugging/index.md if you want a reference.

Verify before pushing. this is the exact command CI runs, it should exit clean with no warnings:


pip install -r docs/requirements.txt
python -m sphinx docs/source docs/build/html -W --keep-going

**Comment by gsmoon97** [comment]
Hi @kaoutar55, thank you for the suggestion. I have added the new toctree in `docs/source/user_guide/examples/index.md` to resolve the build docs CI check fail. I have also verified the fix using sphinx. Thank you very much!

**Review by kaoutar55 — APPROVED** [review]
Thanks @gsmoon97 looks good. 
@mudhakar the latest commits addressed the concern you had. Please re-review and approve. Thanks!

**Review by mudhakar — APPROVED** [review]
